### PR TITLE
Validate that reward functions return one reward per completion

### LINKED ISSUE #6533: GRPO silently accepts a reward function that returns the wrong number of rewards
### Description

`GRPOTrainer` does not check that a custom reward function returns one reward per completion. In `_generate_and_score_completions`, the reward output is assigned straight into the rewards tensor:

```python
rewards_per_func[:, i] = torch.tensor(output_reward_func, dtype=torch.float32, device=device)
```

Because `rewards_per_func[:, i]` has shape `(num_completions,)`, a reward function that returns a single value produces a `(1,)` tensor that PyTorch **silently broadcasts** to every completion. So a broken reward function corrupts training with no warning. A wrong length that is neither 1 nor `num_completions` instead fails later with an opaque tensor shape error rather than a clear message.

### Reproduction

```python
from datasets import load_dataset
from trl import GRPOConfig, GRPOTrainer

dataset = load_dataset("trl-internal-testing/zen", "standard_prompt_only", split="train")

def bad_reward(completions, **kwargs):
    return [1.0]  # returns 1 reward regardless of the number of completions

args = GRPOConfig(
    output_dir="out", per_device_train_batch_size=3, num_generations=3,
    max_completion_length=8, report_to="none",
)
trainer = GRPOTrainer(
    model="trl-internal-testing/tiny-Qwen2ForCausalLM-2.5",
    reward_funcs=bad_reward, args=args, train_dataset=dataset,
)
trainer.train()
```

**Observed:** training runs with no error and logs the single value broadcast as the reward for every completion (`reward: 1`, `reward_std: 0`).

**Expected:** a clear error telling the user the reward function returned the wrong number of rewards.

### Suggested fix

Add a length check in both the synchronous and async reward paths that raises a `ValueError` naming the reward function and the expected count, before the tensor assignment. Happy to open a PR, I have it ready.

### PR DESCRIPTION
**Pull request by gowtham-sai-yadav**
# What does this PR do?

Fixes #6533.

`GRPOTrainer` did not check that a custom reward function returns one reward per completion. The reward output was assigned straight into the rewards tensor:

```python
rewards_per_func[:, i] = torch.tensor(output_reward_func, dtype=torch.float32, device=device)
```

Since `rewards_per_func[:, i]` has shape `(num_completions,)`, a reward function that returns a single value produces a `(1,)` tensor that PyTorch silently broadcasts to every completion, so a broken reward function corrupts training with no warning. A wrong length that is neither 1 nor `num_completions` instead fails later with an opaque tensor shape error.

This adds a length check in both the synchronous and async reward paths that raises a clear `ValueError` naming the reward function and the expected count, right before the tensor assignment. It also adds a regression test (`test_reward_func_wrong_number_of_rewards`).

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case. -> #6533
- [ ] Did you make sure to update the documentation with your changes? (No doc change needed: this only adds a guard on invalid reward-function output.)
- [x] Did you write any new necessary tests? (Added `test_reward_func_wrong_number_of_rewards`.)

## AI writing disclosure

- [ ] No AI usage: the PR was written entirely by a human.
- [x] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

cc @qgallouedec, since this touches the GRPO reward path. Happy to adjust the error wording or add more cases if you'd like.

---

> [!NOTE]
> **Low Risk**
> Small defensive validation on the reward path with clear errors; behavior unchanged for correctly sized reward outputs.
> 
> **Overview**
> Custom reward functions in **GRPO** and **RLOO** must return exactly one reward per prompt–completion pair. This PR adds a **length check** in both trainers (sync and async reward paths) before writing into `rewards_per_func`, raising a **`ValueError`** that names the function and the expected count instead of silently broadcasting a single value or failing later on tensor shape.
> 
> Regression tests **`test_reward_func_wrong_number_of_rewards`** were added for both trainers.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 18d66a37001ab7f0bcc4e91eaf3e1918800e0d1d. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: tests/test_grpo_trainer.py [hunk 1]
```diff
@@ -1663,6 +1663,30 @@ def test_train_with_bias_correction_kl(self, use_liger_kernel):
...
+        dataset = load_dataset("trl-internal-testing/zen", "standard_prompt_only", split="train")
+
+        def wrong_length_reward(completions, **kwargs):
+            return [1.0]  # too few: one reward is expected per completion
+
+        training_args = GRPOConfig(
+            output_dir=self.tmp_dir,
+            per_device_train_batch_size=3,
+            num_generations=3,
+            max_completion_length=8,
```
Comment (qgallouedec) [inline tests/test_grpo_trainer.py]: can you copy the comments that we've for these args from the other tests?

**Review by qgallouedec — CHANGES_REQUESTED** [review]
thanks, can you exactly replicate in RLOO?

File: tests/test_grpo_trainer.py (hunk 1 above)
Comment (gowtham-sai-yadav) [reply tests/test_grpo_trainer.py]: Done, copied the standard arg comments from the other tests.

**Comment by gowtham-sai-yadav** [comment]
> thanks, can you exactly replicate in RLOO?

Also replicated the same check in RLOOTrainer, in both the sync and async reward paths, with a matching regression test in tests/test_rloo_trainer.py. Pushed in 18d66a3.

**Review by qgallouedec — APPROVED** [review]
thanks! useful!

**Comment by gowtham-sai-yadav** [comment]
"Tests with dev dependencies" job failure looks unrelated to this PR. It's the Nemotron "causal_conv1d_fn" error and the gemma4 chat template failures, which also hit the DPO and SFT tests this PR doesn't touch. The reward length tests pass on all the standard matrices.
